### ISSUE TITLE
check_pagefile causing crash on specific server

### ISSUE THREAD

**Issue by MHP47** [mickem/nscp#441]
This is a bit of a strange one, and has me a stumped.
I have NSClient deployed to over 400 machines in my environment, and they all seem to be working fine, except for one. This machine is crashing if the `check_pagefile`  check is called from Nagios (using check_nrpe), or, as discovered in trying to replicate behaviour, if the check is executed twice in succession.

```
C:\Program Files\NSClient++>nscp.exe test
L     client Module: CommandClient
L     client Command:
L     client Extra Query:
L     client Mode: 3
L     client Boot: 1
L     client Load All: 0
L     client Warning module and boot specified only THAT module will be loaded
L     client Arguments:
D       core NSClient++ 0.5.0.62 2016-09-14 x64 Loading settings and logger...
D       core Settings not ready so we cant lookup: base-path
D       core Settings not ready so we cant lookup: exe-path
D   settings Boot.ini found in: C:\Program Files\NSClient++/boot.ini
D   settings Activating: http://<REDACTED>/doc/nsclient.ini
D   settings Creating instance for: http://<REDACTED>/doc/nsclient.ini
D       core Settings not ready so we cant lookup: cache-folder
D       core Settings not ready so we cant lookup: shared-path
D       core Settings not ready so we cant lookup: cache-folder
D       core Settings not ready so we cant lookup: shared-path
D   settings Creating instance for: ini://C:\Program Files\NSClient++/cache\nsclient.ini
D   settings Loading: C:\Program Files\NSClient++/cache\nsclient.ini
D       core NSClient++ 0.5.0.62 2016-09-14 x64 booting...
D       core Booted settings subsystem...
D       core On crash: restart: NSCP
D       core Archiving crash dumps in: C:\Program Files\NSClient++/crash-dumps
D       core booting::loading plugins
D       core Found: CheckDisk
D       core Found: CheckEventLog
D       core Found: CheckExternalScripts
D       core Found: CheckHelpers
D       core Found: CheckNSCP
D       core Found: CheckSystem
D       core Found: NRPEServer
D       core C:\Program Files\NSClient++/modules\CheckDisk.dll.dll
D       core adding C:\Program Files\NSClient++/modules\CheckDisk.dll
D       core C:\Program Files\NSClient++/modules\CheckEventLog.dll.dll
D       core adding C:\Program Files\NSClient++/modules\CheckEventLog.dll
D       core C:\Program Files\NSClient++/modules\CheckExternalScripts.dll.dll
D       core adding C:\Program Files\NSClient++/modules\CheckExternalScripts.dll
D       core C:\Program Files\NSClient++/modules\CheckHelpers.dll.dll
D       core adding C:\Program Files\NSClient++/modules\CheckHelpers.dll
D       core C:\Program Files\NSClient++/modules\CheckNSCP.dll.dll
D       core adding C:\Program Files\NSClient++/modules\CheckNSCP.dll
D       core C:\Program Files\NSClient++/modules\CheckSystem.dll.dll
D       core adding C:\Program Files\NSClient++/modules\CheckSystem.dll
D       core C:\Program Files\NSClient++/modules\NRPEServer.dll.dll
D       core adding C:\Program Files\NSClient++/modules\NRPEServer.dll
D       core Loading plugin: CheckDisk
D       core Loading plugin: CheckEventLog
D       core Loading plugin: CheckExternalScripts
D ext-script No aliases found (adding default)
D       core Loading plugin: CheckHelpers
D       core Loading plugin: CheckNSCP
D check_nscp Crash folder is: C:\Program Files\NSClient++/crash-dumps
D       core Loading plugin: CheckSystem
D       core Loading plugin: NRPEServer
D       nrpe Allowed hosts definition: <REDACTED>
D       nrpe Server config: address: :5666, ssl enabled: none, no certificate, dh: C:\Program Files\NSClient++/security/nrpe_dh_512.pem, ciphers: ADH, ca: C:\Program Files\NSClient++/security/ca.pem, options: no-sslv2,no-sslv3
D  w32system Loading counter: disk_queue_length_0 C: = \\<REDACTED>\PhysicalDisk(0 C:)\% Disk Time
D  w32system Loading counter: disk_queue_length_1 H: = \\<REDACTED>\PhysicalDisk(1 H:)\% Disk Time
D  w32system Loading counter: disk_queue_length_2 I: = \\<REDACTED>\PhysicalDisk(2 I:)\% Disk Time
D  w32system Loading counter: disk_queue_length_3 D: = \\<REDACTED>\PhysicalDisk(3 D:)\% Disk Time
D  w32system Loading counter: disk_queue_length_4 E: = \\<REDACTED>\PhysicalDisk(4 E:)\% Disk Time
D  w32system Loading counter: disk_queue_length_5 F: = \\<REDACTED>\PhysicalDisk(5 F:)\% Disk Time
D  w32system Loading counter: disk_queue_length_6 G: = \\<REDACTED>\PhysicalDisk(6 G:)\% Disk Time
D  w32system Loading counter: disk_queue_length_7 J: = \\<REDACTED>\PhysicalDisk(7 J:)\% Disk Time
D  w32system Loading counter: disk_queue_length_8 K: = \\<REDACTED>\PhysicalDisk(8 K:)\% Disk Time
D  w32system Loading counter: disk_queue_length_9 L: = \\<REDACTED>\PhysicalDisk(9 L:)\% Disk Time
D  w32system Loading counter: disk_queue_length_10 M: = \\<REDACTED>\PhysicalDisk(10 M:)\% Disk Time
D  w32system Loading counter: disk_queue_length_11 N: = \\<REDACTED>\PhysicalDisk(11 N:)\% Disk Time
D  w32system Loading counter: disk_queue_length_12 O: = \\<REDACTED>\PhysicalDisk(12 O:)\% Disk Time
D  w32system Loading counter: disk_queue_length_13 P: = \\<REDACTED>\PhysicalDisk(13 P:)\% Disk Time
D  w32system Loading counter: disk_queue_length_14 Q: = \\<REDACTED>\PhysicalDisk(14 Q:)\% Disk Time
D  w32system Loading counter: disk_queue_length_15 R: = \\<REDACTED>\PhysicalDisk(15 R:)\% Disk Time
D  w32system Loading counter: disk_queue_length_16 S: = \\<REDACTED>\PhysicalDisk(16 S:)\% Disk Time
D  w32system Loading counter: disk_queue_length__Total = \\<REDACTED>\PhysicalDisk(_Total)\% Disk Time
D       nrpe Binding to: [::]:5666(ipv6)
D       nrpe Attempting to bind to: [::]:5666(ipv6)
D       nrpe Binding to: 0.0.0.0:5666(ipv4), reopen: true, reuse: true
D       nrpe Attempting to bind to: 0.0.0.0:5666(ipv4)
D       core NSClient++ - 0.5.0.62 2016-09-14 Started!
D       core C:\Program Files\NSClient++/modules\CommandClient.dll.dll
D       core adding C:\Program Files\NSClient++/modules\CommandClient.dll
D       core Loading plugin: CommandClient...
D        cli Enter command to execute, help for help or exit to exit...
check_pagefile "warning=used > 90" "critical=used > 95"
L        cli CRITICAL: CRITICAL: ??? 148.688GB (0B), total 148.688GB (0B)
L        cli  Performance data: '???'=148.68777GB;0;0;0;0 'total'=148.68777GB;0;0;0;0
check_pagefile "warning=used > 90" "critical=used > 95"
INF: Detected crash...
INF: Running: "C:\Program Files\NSClient++\reporter.exe" "archive" "C:\Users\<REDACTED>\AppData\Local\Temp\4\\b644b36a-e3e0-4ed2-8196-897efc18d2ea.dmp" "NSClient++" "0.5.0.62" "2016-09-14" "C:\Program Files\NSClient++/crash-dumps"
ERR: Service not started, not restarting...

D        cli Enter command to execute, help for help or exit to exit...
check_pagefile "warning=used > 90" "critical=used > 95"
L        cli CRITICAL: CRITICAL: ??? 182.437GB (0B), total 182.437GB (0B)
L        cli  Performance data: '???'=182.43734GB;0;0;0;0 'total'=182.43734GB;0;0;0;0
check_cpu "warning=load > 90" "critical=load > 95"
L        cli OK: OK: CPU load is ok.
L        cli  Performance data: 'total 5m'=0%;90;95 'total 1m'=1%;90;95 'total 5s'=11%;90;95
check_pagefile "warning=used > 90" "critical=used > 95"
L        cli CRITICAL: CRITICAL: ??? 182.437GB (0B), total 182.437GB (0B)
L        cli  Performance data: '???'=182.43734GB;0;0;0;0 'total'=182.43734GB;0;0;0;0
check_pagefile "warning=used > 90" "critical=used > 95"
INF: Detected crash...
INF: Running: "C:\Program Files\NSClient++\reporter.exe" "archive" "C:\Users\<REDACTED>\AppData\Local\Temp\4\\130e3c92-d3f0-4ee6-b2ea-2dada195b1ea.dmp" "NSClient++" "0.5.0.62" "2016-09-14" "C:\Program Files\NSClient++/crash-dumps"
ERR: Service not started, not restarting...

D        cli Enter command to execute, help for help or exit to exit...
D       nrpe Accepting connection from: 10.224.176.205, count=1
D       nrpe Accepting connection from: 10.224.176.205, count=1
D       nrpe Accepting connection from: 10.224.176.205, count=1
D       nrpe Accepting connection from: 10.224.176.205, count=1
D       nrpe Accepting connection from: 10.224.176.205, count=1
D       nrpe Accepting connection from: 10.224.176.205, count=1
INF: Detected crash...
INF: Running: "C:\Program Files\NSClient++\reporter.exe" "archive" "C:\Users\<REDACTED>\AppData\Local\Temp\4\\9bce0809-bb37-480f-a72f-b09cdb5cd693.dmp" "NSClient++" "0.5.0.62" "2016-09-14" "C:\Program Files\NSClient++/crash-dumps"
ERR: Service not started, not restarting...

>/usr/local/nagios/libexec/check_nrpe -H <REDACTED> -c check_cpu -a "warning=load > 90" "critical=load > 95"
OK: CPU load is ok.|'total 5m'=0%;90;95 'total 1m'=1%;90;95 'total 5s'=12%;90;95
>/usr/local/nagios/libexec/check_nrpe -H <REDACTED> -c check_memory -a "warning=used > 90" "critical=used > 95"
OK: committed = 4.115GB, physical = 5.862GB|'committed'=4.11465GB;21.58938;22.78879;0;23.9882 'committed %'=17%;90;95;0;100 'physical'=5.86217GB;21.59098;22.79048;0;23.98997 'physical %'=24%;90;95;0;100
>/usr/local/nagios/libexec/check_nrpe -H <REDACTED> -c check_pagefile -a "warning=used > 90" "critical=used > 95"
CHECK_NRPE: Receive header underflow - only -1 bytes received (4 expected).
```

There's 3 scenarios in the above log.

1. First is running the `check_pagefile` twice in a row. First execution appears to work ok, and the second causes a crash.
2. Second shows running the check successfully, then another check, then back to  `check_pagefile`. It then crashes when run twice in a row. If another check is run in between it seems to be ok (not shown, but did test this multiple times in a row, switching from one check to the other with no crash)
3.  This one shows the connections from Nagios (with check_nrpe). Ran 3 commands, as shown underneath. Crashes as soon as the check_pagefile is executed. 

Ran the exact same tests on a different server with NSClient installed, and it did not report the same symptoms (no crash experienced). There's something about this specific server, but I am unsure where/what to be looking at.

Opened the dump files generated, which contained the below:
```
Exception Code           0xC0000005
Exception Information    The thread tried to read from or write to a virtual address for which it does not have the appropriate access.

OS Version               6.1.7601
```

**Comment by mickem** [comment]
I know there was some issues with check_pagefile which has since been fixed in 0.5.1.8 so you might anna try and see if that resolves the issue: https://nsclient.org/download/0.5.1/#0.5.1.8

**Comment by mickem** [comment]
Please reopen if this can be reproduced with latest builds...
